lib/backend/etcdbk/atomicwrite.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Package etcdbk implements Etcd powered backend
+package etcdbk
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"github.com/gravitational/trace"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+func (b *EtcdBackend) AtomicWrite(ctx context.Context, condacts ...backend.ConditionalAction) (revision string, err error) {
+	if err := backend.ValidateAtomicWrite(condacts); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	var cmps []clientv3.Cmp
+	var ops []clientv3.Op
+	var includesPut bool
+
+	for _, ca := range condacts {
+		key := b.prependPrefix(ca.Key)
+
+		switch ca.Condition.Kind {
+		case backend.KindWhatever:
+			// no comparison to assert
+		case backend.KindExists:
+			cmps = append(cmps, clientv3.Compare(clientv3.CreateRevision(key), "!=", 0))
+		case backend.KindNotExists:
+			cmps = append(cmps, clientv3.Compare(clientv3.CreateRevision(key), "=", 0))
+		case backend.KindRevision:
+			rev, err := fromBackendRevision(ca.Condition.Revision)
+			if err != nil {
+				// malformed revisions are considered a kind of failed condition since they indicate that
+				// the supplied revision did not originate from a preceding read from this backend.
+				return "", trace.Wrap(backend.ErrConditionFailed)
+			}
+
+			cmps = append(cmps, clientv3.Compare(clientv3.CreateRevision(key), "!=", 0))
```

Comment:
Do we need the existence check if we are asserting that the mod revision is a specific value? Is this just so a revision of `"0"` doesn't end up checking for nonexistence? If so maybe we could just prohibit `"0"` as a revision in `etcdbk.fromBackendRevision` instead.